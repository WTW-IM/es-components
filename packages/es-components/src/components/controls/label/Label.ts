import styled from 'styled-components';
import getStyledProp, { ESThemeProps } from '../../util/getStyledProp';

const Label = styled.label`
  color: ${getStyledProp('colors.gray9')};
  cursor: pointer;
  font-family: 'Source Sans Pro', 'Segoe UI', Segoe, Calibri, Tahoma, sans-serif;
  font-size: ${props =>
    getStyledProp('font.labelFontSize', props as ESThemeProps) ||
    getStyledProp('font.baseFontSize', props as ESThemeProps)};
  font-weight: ${props =>
    getStyledProp('font.labelFontWeight', props as ESThemeProps) || 'bold'};
  line-height: ${getStyledProp('font.baseLineHeight')};
  margin-bottom: 5px;
  display: inline-block;

  &[disabled] {
    cursor: not-allowed;
  }

  @media (min-width: ${getStyledProp('screenSize.tablet')}) {
    margin-right: 20px;
  }
`;

/** @component */
export default Label;
