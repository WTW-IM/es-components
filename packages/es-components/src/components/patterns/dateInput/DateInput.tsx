import 'get-root-node-polyfill/implement';
import React, { useEffect, useReducer, useRef, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { isBefore, isAfter } from 'date-fns';

import Day, { DayProps } from './Day';
import Month, { MonthProps } from './Month';
import Year, { YearProps } from './Year';

import isBool from '../../util/isBool';
import noop from '../../util/noop';

const Wrapper = styled.div`
  display: flex;

  && > * {
    margin-right: 2px;
  }

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    max-width: 350px;
  }
`;

type DayNumber = `${'0' | ''}${number}`;
type NonZero = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';
type CleanNumber = `${NonZero}${number | ''}`;

function cleanZeroes(day: DayNumber) {
  let cleanDay: string = day;
  while (cleanDay.charAt(0) === '0') {
    cleanDay = cleanDay.substr(1);
  }
  return cleanDay as CleanNumber;
}

const dateParts = ['day', 'month', 'year'] as const;

export type DatePart = (typeof dateParts)[number];

export type DatePartChangeHandler = (part: DatePart, value: string) => void;

export type DateInputState = {
  [key in DatePart]?: string;
};

type DateAction = {
  type: 'day_updated' | 'month_updated' | 'year_updated';
  value: string;
};

function reducer(state: DateInputState, action: DateAction): DateInputState {
  switch (action.type) {
    case 'day_updated':
      return { ...state, day: action.value };
    case 'month_updated':
      return { ...state, month: action.value };
    case 'year_updated':
      return { ...state, year: action.value };
    default:
      throw new Error();
  }
}

type InputProps = JSXElementProps<'input'>;
export type DateChangeEvent = {
  value?: Date;
  isInRange: boolean;
  rawValues: {
    year: string;
    month: string;
    day: string;
  };
};

export type DateInputProps = Override<
  JSXElementProps<'div'>,
  {
    children: React.ReactNode;
    onChange: (event: DateChangeEvent) => void;
    id?: string;
    maxDate?: Date;
    minDate?: Date;
    onBlur?: InputProps['onBlur'];
    defaultValue?: Date;
    defaultDay?: string;
  }
>;

const getDefaultState = (
  defaultValue?: Date,
  defaultDay = ''
): DateInputState => {
  if (!defaultValue) {
    return {
      day: defaultDay,
      month: '',
      year: ''
    };
  }

  return {
    day: defaultValue.getDate().toString(),
    month: `${defaultValue.getMonth() + 1}`,
    year: defaultValue.getFullYear().toString()
  };
};

const isIterable = (value: unknown): value is Iterable<unknown> => {
  return typeof (value as Iterable<unknown>)[Symbol.iterator] === 'function';
};

function isReactElementChild(
  child: React.ReactNode
): child is React.ReactElement {
  return !(
    typeof child === 'string' ||
    typeof child === 'number' ||
    isBool(child) ||
    isIterable(child) ||
    Array.isArray(child)
  );
}

function isDatePartChild(
  child: React.ReactNode
): child is React.ReactElement<DayProps | MonthProps | YearProps> {
  return (
    isReactElementChild(child) &&
    (child.type === Day || child.type === Month || child.type === Year)
  );
}

const DateInput = React.forwardRef<HTMLDivElement, DateInputProps>(
  function ForwardedDateInput(
    {
      children,
      defaultValue,
      defaultDay = '',
      id,
      maxDate,
      minDate,
      onBlur = noop,
      onChange,
      ...props
    },
    ref
  ) {
    const [state, dispatch] = useReducer<typeof reducer>(
      reducer,
      getDefaultState(defaultValue, defaultDay)
    );

    const hasDayElement = useRef(false);

    useEffect(() => {
      React.Children.forEach(children, child => {
        if (!child) return;

        if (!isReactElementChild(child)) return;

        if (child.type === Month && !state.month) {
          if ((child.props as MonthProps).selectOptionText) {
            dispatch({ type: 'month_updated', value: 'none' });
          }
        }
        if (child.type === Day) {
          hasDayElement.current = true;
        }
      });
    }, [children, state.month]);

    const createDate = useCallback(
      (year?: string, month?: string, day?: string): DateChangeEvent => {
        const cleanDay = day ? cleanZeroes(day.toString() as DayNumber) : day;
        const date = new Date(`${year}/${month}/${cleanDay}`);
        const dateIsValid =
          month !== 'none' &&
          year?.length === 4 &&
          (hasDayElement.current
            ? date.getDate().toString() === cleanDay
            : true) &&
          (month
            ? (date.getMonth() + 1).toString() === month.toString()
            : true) &&
          date.getFullYear().toString() === year.toString();

        const isInRange =
          dateIsValid &&
          (minDate ? isAfter(date, minDate) : true) &&
          (maxDate ? isBefore(date, maxDate) : true);
        return {
          value: dateIsValid ? date : undefined,
          isInRange,
          rawValues: {
            year: year?.toString() || '',
            month: month === 'none' ? '' : month?.toString() || '',
            day: day?.toString() || ''
          }
        };
      },
      [maxDate, minDate]
    );

    const onChangeDatePart = useCallback(
      (datePart: DatePart, value: string) => {
        const { day, month, year } = state;
        switch (datePart) {
          case 'day':
            onChange(createDate(year, month, value));
            dispatch({
              type: 'day_updated',
              value
            });
            return;
          case 'month':
            onChange(createDate(year, value, day));
            dispatch({ type: 'month_updated', value });
            return;
          case 'year':
            onChange(createDate(value, month, day));
            dispatch({
              type: 'year_updated',
              value
            });
            return;
          default:
            throw new Error();
        }
      },
      [createDate, onChange, state]
    );

    const onBlurComponent = useCallback<
      React.FocusEventHandler<HTMLInputElement>
    >(
      event => {
        const target = event.currentTarget;
        setTimeout(() => {
          if (
            !target.contains(
              (target.getRootNode() as unknown as DocumentOrShadowRoot)
                .activeElement || null
            )
          ) {
            onBlur(event);
          }
        }, 0);
      },
      [onBlur]
    );

    let setId = false;

    return (
      <Wrapper ref={ref} tabIndex={-1} {...props} onBlur={onBlurComponent}>
        {React.Children.map(children, child =>
          isDatePartChild(child)
            ? React.cloneElement(child, {
                id:
                  !setId && id
                    ? (setId = true) && id
                    : (child.props as HTMLElementProps).id,
                onChange: onChangeDatePart,
                date: createDate(state.year, state.month, state.day).value
              })
            : child
        )}
      </Wrapper>
    );
  }
);

type DateInputType = typeof DateInput & {
  Day: typeof Day;
  Month: typeof Month;
  Year: typeof Year;
};

DateInput.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node.isRequired,
  /** set the minimum valid date */
  maxDate: PropTypes.instanceOf(Date),
  /** set the maximum valid date */
  minDate: PropTypes.instanceOf(Date),
  /** returns an object (value, isInRange) */
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  /** set the Date value of the component */
  defaultValue: PropTypes.instanceOf(Date),
  /** Set the default day of the month */
  defaultDay: PropTypes.string
};

DateInput.defaultProps = {
  id: undefined,
  maxDate: undefined,
  minDate: undefined,
  defaultValue: undefined,
  defaultDay: '',
  onBlur: noop
};

const setParts = (Input: DateInputType) => {
  Input.Day = Day;
  Input.Month = Month;
  Input.Year = Year;
};

setParts(DateInput as DateInputType);

export default DateInput as DateInputType;
