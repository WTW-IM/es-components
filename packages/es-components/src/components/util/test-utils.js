// eslint-env jest
import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function ThemeComponent(props) {
  return <ThemeProvider theme={viaTheme} {...props} />;
}

export function renderWithTheme(component) {
  return render(<ThemeComponent>{component}</ThemeComponent>);
}
