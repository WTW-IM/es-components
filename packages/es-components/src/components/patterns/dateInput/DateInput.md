<a href="https://8lf1uv.axshare.com/#id=7zaoiz&p=multi_field_error&dp=0&g=1" target="blank"><div style="color:#cc0000;text-transform:uppercase;margin-bottom:1em;">View Error Pattern Examples</div></a>

Using the `onChange` callback returns an object:

```html
{ value: [Date | undefined] isInRange: [true | false] }
```

The `value` is a Date created from the combined inputs. The `value` is `undefined` when the date is invalid.
`isInRange` is provided for validation help.

`DateInput.Month` also accepts an array of month names (`monthName` prop) for localization support. You may also
provide `selectOptionText` to define a default "Please select a Month"-style option.

```jsx
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import AdditionalHelp from '../../controls/AdditionalHelp';

function DateInputExample() {
  const options = {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState('default');

  function handleChange(date) {
    console.log(date);
    setMyDate(date.value);
    setValidation('default');
  }

  function handleOnBlur() {
    setValidation(myDate ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <Label htmlFor="someDate">Enter a Date</Label>
      <DateInput
        id="someDate"
        name="someDate"
        onChange={handleChange}
        onBlur={handleOnBlur}
      >
        <DateInput.Month
          monthNames={['01 - Jan', '02 - Feb', '03 - Mar']}
          selectOptionText="Select a month"
        />
        <DateInput.Day />
        <DateInput.Year />
      </DateInput>
      <AdditionalHelp>
        {myDate && (
          <span>
            You entered: {myDate.toLocaleDateString('en-US', options)}
          </span>
        )}
      </AdditionalHelp>
    </Control>
  );
}

<DateInputExample />;
```

You can set the value of the component with `defaultValue`. When populating this value from
an API call or other source via a `useEffect`, the `defaultValue` may not update correctly.
In this case, provide `DateInput` with a `key` prop that changes along with the `defaultValue`
to force the component to re-render with the new value.

```jsx
<DateInput defaultValue={new Date()} onChange={() => {}}>
  <DateInput.Month name="month" />
  <DateInput.Day name="day" />
  <DateInput.Year name="year" />
</DateInput>
```

### Min and Max dates

Use `minDate` and `maxDate` to set restrictions on valid dates.

```jsx
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import AdditionalHelp from '../../controls/AdditionalHelp';

function DateInputExample() {
  const options = {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState('default');

  function handleChange(date) {
    console.log(date);
    setMyDate(date);
    setValidation('default');
  }

  function handleOnBlur() {
    setValidation(myDate && myDate.isInRange ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <Label htmlFor="demoDate">Enter a Date between 1900 and 2020</Label>
      <DateInput
        id="demoDate"
        onChange={handleChange}
        onBlur={handleOnBlur}
        minDate={new Date('1900/1/1')}
        maxDate={new Date('2020/1/1')}
      >
        <DateInput.Month />
        <DateInput.Day />
        <DateInput.Year />
      </DateInput>
      <AdditionalHelp>
        <span>
          {myDate && myDate.isInRange
            ? `You entered: ${myDate.value.toLocaleDateString(
                'en-US',
                options
              )}`
            : validation !== 'default' &&
              'The date you entered is outside the valid range.'}
        </span>
      </AdditionalHelp>
    </Control>
  );
}

<DateInputExample />;
```

### Display

You may omit Day, but Month and Year is required.

```jsx
<DateInput id="anotherDate" onChange={console.log}>
  <DateInput.Month />
  <DateInput.Year />
</DateInput>
```

Month and Year validated with no default selected.

```jsx
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import AdditionalHelp from '../../controls/AdditionalHelp';

function DateInputExample() {
  const options = {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState('default');

  function handleChange(date) {
    console.log(date);
    setMyDate(date);
    setValidation('default');
  }

  function handleOnBlur() {
    setValidation(myDate && myDate.value ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <DateInput
        id="demoDate"
        onChange={handleChange}
        onBlur={handleOnBlur}
        minDate={new Date('2000/1/1')}
        maxDate={new Date('2040/1/1')}
      >
        <DateInput.Month selectOptionText="Select a month" />
        <DateInput.Year />
      </DateInput>
      <AdditionalHelp>
        <span>
          {myDate && myDate.isInRange
            ? `You entered: ${myDate.value.toLocaleDateString(
                'en-US',
                options
              )}`
            : validation !== 'default' && 'Please enter a valid date.'}
        </span>
      </AdditionalHelp>
    </Control>
  );
}
<DateInputExample />;
```

Set default day if no day input is provided and no default value is provided.

```jsx
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import AdditionalHelp from '../../controls/AdditionalHelp';

function DateInputExample() {
  const options = {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState('default');

  function handleChange(date) {
    console.log(date);
    setMyDate(date);
    setValidation('default');
  }

  function handleOnBlur() {
    setValidation(myDate && myDate.value ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <DateInput
        id="demoDate"
        onChange={handleChange}
        onBlur={handleOnBlur}
        minDate={new Date('2000/1/1')}
        maxDate={new Date('2040/1/1')}
        defaultDay="1"
      >
        <DateInput.Month selectOptionText="Select a month" />
        <DateInput.Year />
      </DateInput>
      <AdditionalHelp>
        <span>
          {myDate && myDate.isInRange
            ? `You entered: ${myDate.value.toLocaleDateString(
                'en-US',
                options
              )}`
            : validation !== 'default' && 'Please enter a valid date.'}
        </span>
      </AdditionalHelp>
    </Control>
  );
}
<DateInputExample />;
```

Year, Month, and Day will display in the order you provide them.

```jsx
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';

<Control>
  <Label htmlFor="orderDate">Enter a Date</Label>
  <DateInput id="orderDate" defaultValue={new Date()} onChange={() => {}}>
    <DateInput.Year />
    <DateInput.Month />
    <DateInput.Day />
  </DateInput>
</Control>;
```
