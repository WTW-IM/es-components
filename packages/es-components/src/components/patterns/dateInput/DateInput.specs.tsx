import React from 'react';
import { waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from '../../util/test-utils';
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import DateInput, { DateInputProps } from './DateInput';

const typeFresh = async (...args: Parameters<typeof userEvent.type>) => {
  await userEvent.clear(args[0]);
  return await userEvent.type(...args);
};

const TestDateInput = (props: DateInputProps) => (
  <Control>
    <DateInput {...props} />
  </Control>
);

it('executes the onChange function when text is changed', async () => {
  const onChange = jest.fn();

  renderWithTheme(
    <TestDateInput onChange={onChange}>
      <DateInput.Month />
      <DateInput.Day data-testid="test-day" />
      <DateInput.Year />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('test-day'), '21');

  expect(onChange).toHaveBeenCalled();
});

it('correctly adds id to first DatePart element', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1)
  };

  renderWithTheme(
    <Control>
      <Label htmlFor="test">Label Text</Label>
      <DateInput id="test" {...props}>
        <Label>Month</Label>
        <DateInput.Month data-testid="month" />
        <Label>Day</Label>
        <DateInput.Day />
        <Label>Year</Label>
        <DateInput.Year />
      </DateInput>
    </Control>
  );

  const targetInput = await screen.findByLabelText('Label Text');
  expect(targetInput).toHaveAttribute('id', 'test');
  expect(targetInput).toHaveAttribute('data-testid', 'month');
});

it('executes handleOnBlur when focus is lost', async () => {
  const onBlur = jest.fn();
  const onChange = jest.fn();

  renderWithTheme(
    <TestDateInput onChange={onChange} onBlur={onBlur}>
      <DateInput.Month />
      <DateInput.Day />
      <DateInput.Year data-testid="test-year" />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('test-year'), '2025');
  await userEvent.tab();

  await waitFor(() => {
    expect(onBlur).toHaveBeenCalled();
  });
});

it('displays the value properly when defaultValue is set', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1)
  };

  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day />
      <DateInput.Year data-testid="test-year" />
    </TestDateInput>
  );

  expect(await screen.findByDisplayValue('1')).not.toBeNull();
  expect(await screen.findByDisplayValue('March')).not.toBeNull();
  expect(await screen.findByDisplayValue('2019')).not.toBeNull();
});

it('hides the Day input when Day is excluded', () => {
  const props = {
    onChange: jest.fn()
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Year />
    </TestDateInput>
  );

  expect(
    screen.queryByRole('spinbutton', { name: /day/i })
  ).not.toBeInTheDocument();
});

it('displays monthNames provided by the monthNames prop', async () => {
  const monthNames = ['Jan', 'Feb', 'Mar'];
  const extraMonths = [
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ];
  const props = {
    onChange: jest.fn()
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month name="month" monthNames={monthNames} />
      <DateInput.Day name="day" />
      <DateInput.Year name="year" />
    </TestDateInput>
  );

  for (const month of monthNames) {
    expect(
      await screen.findByRole('option', { name: new RegExp(month) })
    ).toBeInTheDocument();
  }

  for (const month of extraMonths) {
    expect(
      screen.queryByRole('option', { name: new RegExp(month) })
    ).not.toBeInTheDocument();
  }
});

it('sets date invalid when before minDate', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1),
    minDate: new Date(2000, 1, 1)
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day />
      <DateInput.Year data-testid="year-input" />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('year-input'), '1875');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    rawValues: {
      day: '1',
      month: '3',
      year: '1875'
    },
    value: new Date(1875, 2, 1)
  });
});

it('sets date invalid when after maxDate', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1),
    maxDate: new Date(2000, 1, 1)
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day />
      <DateInput.Year data-testid="year-input" />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('year-input'), '2001');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    rawValues: {
      day: '1',
      month: '3',
      year: '2001'
    },
    value: new Date(2001, 2, 1)
  });
});

it('returns undefined when bad date is entered', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 1, 15) // feb
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day data-testid="day-input" />
      <DateInput.Year />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('day-input'), '31');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    rawValues: {
      day: '31',
      month: '2',
      year: '2019'
    },
    value: undefined
  });
});

it('returns a date when valid', async () => {
  const props = {
    onChange: jest.fn(),
    minDate: new Date(2000, 1, 1),
    maxDate: new Date(2020, 1, 1),
    defaultValue: new Date(2019, 1, 15) // feb
  };
  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day data-testid="day-input" />
      <DateInput.Year />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('day-input'), '16');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: true,
    rawValues: {
      day: '16',
      month: '2',
      year: '2019'
    },
    value: new Date(2019, 1, 16)
  });
});

it('returns undefined when Day is present but not defined', async () => {
  const props = {
    onChange: jest.fn()
  };

  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day />
      <DateInput.Year data-testid="year-input" />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('year-input'), '1983');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    rawValues: {
      day: '',
      month: '',
      year: '1983'
    },
    value: undefined
  });
});

it('returns date when Day is padded with 0', async () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 0, 1)
  };

  renderWithTheme(
    <TestDateInput {...props}>
      <DateInput.Month />
      <DateInput.Day data-testid="day-input" />
      <DateInput.Year data-testid="year-input" />
    </TestDateInput>
  );

  await typeFresh(await screen.findByTestId('day-input'), '02');
  await typeFresh(await screen.findByTestId('year-input'), '1983');

  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: true,
    rawValues: {
      day: '2', // jsdom clears the leading 0 for input type 'number'
      month: '1',
      year: '1983'
    },
    value: new Date(1983, 0, 2)
  });
});
