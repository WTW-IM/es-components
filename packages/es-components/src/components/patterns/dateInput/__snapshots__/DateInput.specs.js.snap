// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays monthNames provided by the monthNames prop 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c0 >[role="group"] {
  margin-bottom: 0;
}

.c0 select {
  border-color: #c5c5c5;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
}

.c0 input:focus,
.c0 select:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
}

.c1 {
  color: #444;
  cursor: pointer;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 18px;
  font-weight: bold;
  line-height: 1.428;
  margin-bottom: 5px;
  display: inline-block;
}

.c1[disabled] {
  cursor: not-allowed;
}

.c5 {
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
  box-sizing: border-box;
  color: #000;
  font-size: 18px;
  font-face: "Source Sans Pro","Segoe UI",Segoe,Calibri,Tahoma,sans-serif;
  font-weight: normal;
  height: 39px;
  line-height: 1.428;
  min-width: 0;
  padding: 6px 12px;
  -webkit-transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  width: 100%;
}

.c5:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
  outline: 0;
}

.c5:disabled {
  background-color: #ededed;
  cursor: not-allowed;
}

.c5:-moz-read-only {
  background-color: #ededed;
  cursor: text;
}

.c5:read-only {
  background-color: #ededed;
  cursor: text;
}

.c6 {
  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  appearance: textfield;
  -webkit-flex: 1 0 35px;
  -ms-flex: 1 0 35px;
  flex: 1 0 35px;
}

.c6::-webkit-inner-spin-button,
.c6::-webkit-outer-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
}

.c3 {
  background-color: #fff;
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
  box-sizing: border-box;
  color: #000;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 18px;
  font-weight: normal;
  height: 39px;
  min-width: 100px;
  padding: 6px 12px;
  -webkit-transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
}

.c3:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
  outline: 0;
}

.c3:disabled {
  background-color: #ededed;
  cursor: not-allowed;
}

.c4 {
  -webkit-flex: 3 1 100px;
  -ms-flex: 3 1 100px;
  flex: 3 1 100px;
}

.c7 {
  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  appearance: textfield;
  -webkit-flex: 2 0 40px;
  -ms-flex: 2 0 40px;
  flex: 2 0 40px;
}

.c7::-webkit-inner-spin-button,
.c7::-webkit-outer-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2.c2 > * {
  margin-right: 2px;
}

@media (min-width:768px) {
  .c1 {
    margin-right: 20px;
  }
}

@media (min-width:768px) {
  .c2 {
    max-width: 350px;
  }
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      Text
    </label>
    <div
      class="c2"
      data-testid="dateinput"
      tabindex="-1"
    >
      <select
        aria-label="Select Month"
        class="c3 c4"
        id="test"
        name="month"
      >
        <option
          value="1"
        >
          Jan
        </option>
        <option
          value="2"
        >
          Feb
        </option>
        <option
          value="3"
        >
          Mar
        </option>
      </select>
      <input
        aria-label="Day"
        class="c5 c6"
        max="31"
        min="1"
        name="day"
        pattern="[0-9]*"
        placeholder="Day"
        type="number"
        value=""
      />
      <input
        aria-label="Year"
        class="c5 c7"
        name="year"
        pattern="[0-9]*"
        placeholder="Year"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`hides the Day input when Day is excluded 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c0 >[role="group"] {
  margin-bottom: 0;
}

.c0 select {
  border-color: #c5c5c5;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
}

.c0 input:focus,
.c0 select:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
}

.c1 {
  color: #444;
  cursor: pointer;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 18px;
  font-weight: bold;
  line-height: 1.428;
  margin-bottom: 5px;
  display: inline-block;
}

.c1[disabled] {
  cursor: not-allowed;
}

.c5 {
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
  box-sizing: border-box;
  color: #000;
  font-size: 18px;
  font-face: "Source Sans Pro","Segoe UI",Segoe,Calibri,Tahoma,sans-serif;
  font-weight: normal;
  height: 39px;
  line-height: 1.428;
  min-width: 0;
  padding: 6px 12px;
  -webkit-transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  width: 100%;
}

.c5:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
  outline: 0;
}

.c5:disabled {
  background-color: #ededed;
  cursor: not-allowed;
}

.c5:-moz-read-only {
  background-color: #ededed;
  cursor: text;
}

.c5:read-only {
  background-color: #ededed;
  cursor: text;
}

.c3 {
  background-color: #fff;
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
  box-sizing: border-box;
  color: #000;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 18px;
  font-weight: normal;
  height: 39px;
  min-width: 100px;
  padding: 6px 12px;
  -webkit-transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
  transition: border-color ease-in-out 0.15s,box-shadow ease-in-out 0.15s;
}

.c3:focus {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6);
  outline: 0;
}

.c3:disabled {
  background-color: #ededed;
  cursor: not-allowed;
}

.c4 {
  -webkit-flex: 3 1 100px;
  -ms-flex: 3 1 100px;
  flex: 3 1 100px;
}

.c6 {
  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  appearance: textfield;
  -webkit-flex: 2 0 40px;
  -ms-flex: 2 0 40px;
  flex: 2 0 40px;
}

.c6::-webkit-inner-spin-button,
.c6::-webkit-outer-spin-button {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2.c2 > * {
  margin-right: 2px;
}

@media (min-width:768px) {
  .c1 {
    margin-right: 20px;
  }
}

@media (min-width:768px) {
  .c2 {
    max-width: 350px;
  }
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      Text
    </label>
    <div
      class="c2"
      data-testid="dateinput"
      tabindex="-1"
    >
      <select
        aria-label="Select Month"
        class="c3 c4"
        id="test"
      >
        <option
          value="1"
        >
          January
        </option>
        <option
          value="2"
        >
          February
        </option>
        <option
          value="3"
        >
          March
        </option>
        <option
          value="4"
        >
          April
        </option>
        <option
          value="5"
        >
          May
        </option>
        <option
          value="6"
        >
          June
        </option>
        <option
          value="7"
        >
          July
        </option>
        <option
          value="8"
        >
          August
        </option>
        <option
          value="9"
        >
          September
        </option>
        <option
          value="10"
        >
          October
        </option>
        <option
          value="11"
        >
          November
        </option>
        <option
          value="12"
        >
          December
        </option>
      </select>
      <input
        aria-label="Year"
        class="c5 c6"
        pattern="[0-9]*"
        placeholder="Year"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;
